File: frontend/eskap_app/lib/models/escapeGame.dart
import 'package:equatable/equatable.dart';
import 'package:eskap_app/models/review.dart';

class EscapeGame extends Equatable {
  final int id;
  final String name;
  final String difficulty;

  final double minprice;
  final double maxprice;

  final String imgurl;
  final String websiteurl;

  final String description;

  final int minplayer;
  final int maxplayer;

  final int roomnumber;

  // Address
  final int number;
  final String street;
  final String city;
  final String country;
  final double latitude;
  final double longitude;
  //
  final List<String> themes;
  final List<Review> reviews;

  final bool official;

  final bool isFav;
  final bool isDone;

  EscapeGame({
    this.id,
    this.name,
    this.difficulty,
    this.minprice,
    this.maxprice,
    this.imgurl,
    this.websiteurl,
    this.description,
    this.minplayer,
    this.maxplayer,
    this.roomnumber,
    this.number,
    this.street,
    this.city,
    this.country,
    this.latitude,
    this.longitude,
    this.themes,
    this.reviews,
    this.official,
    this.isFav,
    this.isDone,
  });

  @override
  String toString() {
    return 'EscapeGame{name : $name , adresse: $addressToString(), lat: $latitude, Lng: $longitude, id: $id}';
  }

  String addressToString() {
    if (street == "" && number == 0) return 'Adresse : $city $country';
    return 'Adresse : $number $street $city $country';
  }

  Map<String, dynamic> toJson() => {
        "name": name,
        "difficulty": difficulty,
        "minprice": minprice,
        "maxprice": maxprice,
        "imgurl": imgurl,
        "websiteurl": websiteurl,
        "description": description,
        "minplayer": minplayer,
        "maxplayer": maxplayer,
        "roomnumber": roomnumber,
        "number": number,
        "street": street,
        "city": city,
        "country": country,
        "latitude": latitude,
        "longitude": longitude,
        "themes": themes,
        "reviews": reviews,
        "official": official,
      };

  static EscapeGame fromJson(eskap, userId, bool isFav, bool isDone) {
    return EscapeGame(
        id: eskap['id'],
        name: eskap['name'],
        difficulty: eskap['difficulty'],
        minprice: eskap['minprice'],
        maxprice: eskap['maxprice'],
        imgurl: eskap['imgurl'],
        websiteurl: eskap['websiteurl'],
        description: eskap['description'],
        minplayer: eskap['minplayer'],
        maxplayer: eskap['maxplayer'],
        roomnumber: eskap['roomnumber'],
        number: eskap['number'],
        street: eskap['street'],
        city: eskap['city'],
        country: eskap['country'],
        latitude: eskap['latitude'],
        longitude: eskap['longitude'],
        themes: themesFromJson(eskap['themes']),
        reviews: reviewsFromJson(eskap['reviews'], userId),
        official: eskap['official'] ?? false,
        isFav: isFav,
        isDone: isDone);
  }

  static EscapeGame updateEscapeFromPrevious(EscapeGame eg,
      {bool isFav, bool isDone, Review review}) {
    return EscapeGame(
      id: eg.id,
      name: eg.name,
      difficulty: eg.difficulty,
      minprice: eg.minprice,
      imgurl: eg.imgurl,
      description: eg.description,
      number: eg.number,
      street: eg.street,
      city: eg.city,
      country: eg.country,
      latitude: eg.latitude,
      longitude: eg.longitude,
      themes: eg.themes,
      reviews: review != null ? [...eg.reviews, review] : eg.reviews,
      official: eg.official,
      isFav: isFav == null ? eg.isFav : isFav,
      isDone: isDone == null ? eg.isDone : isDone,
    );
  }

  static List<String> themesFromJson(themes) {
    List<String> res = [];
    for (var theme in themes) {
      res.add(theme.toString());
    }
    return res;
  }

  String themesToString() {
    return themes.join(' ,');
  }

  static List<Review> reviewsFromJson(reviews, userId) {
    List<Review> res = [];
    for (var review in reviews) {
      res.add(Review.fromJson(review, userId));
    }
    return res;
  }

  double averageRate() {
    double res = 0;
    if (reviews.length == 0) return res;
    for (var review in reviews) res += review.rate;
    return res / (reviews.length);
  }

  @override
  List<Object> get props => [
        id,
        name,
        difficulty,
        minprice,
        maxprice,
        imgurl,
        websiteurl,
        description,
        minplayer,
        maxplayer,
        roomnumber,
        number,
        street,
        city,
        country,
        latitude,
        longitude,
        themes,
        reviews,
        official,
        isFav,
        isDone,
      ];
}

File: backend/escape-game-api/src/main/java/com/ustl/eskap/app/bo/eskap/EscapeGame.java
package com.ustl.eskap.app.bo.eskap;

import javax.persistence.*;
import java.util.List;

@Entity
public class EscapeGame {

    @Id
    @GeneratedValue(strategy= GenerationType.AUTO)
    private int id;

    //EscapeGameInfos
    private String name;
    private String difficulty;
    private double minprice;
    private double maxprice;
    private String imgurl;
    private String websiteurl;
    private String description;
    private int minplayer;
    private int maxplayer;
    private int roomnumber;

    // Address
    private int number;
    private String street;
    private String city;
    private String country;
    private double latitude;
    private double longitude;

    // Collections
    @ElementCollection
    private List<String> themes;

    @ElementCollection
    private List<Review> reviews;

    // Official
    private Boolean isOfficial;

    public EscapeGame() {

    }

    public EscapeGame(int id, String name, String difficulty, double minprice, double maxprice, String imgurl, String websiteurl, String description, int minplayer, int maxplayer, int roomnumber, int number, String street, String city, String country, double latitude, double longitude, List<String> themes, List<Review> reviews, Boolean isOfficial) {
        this.id = id;
        this.name = name;
        this.difficulty = difficulty;
        this.minprice = minprice;
        this.maxprice = maxprice;
        this.imgurl = imgurl;
        this.websiteurl = websiteurl;
        this.description = description;
        this.minplayer = minplayer;
        this.maxplayer = maxplayer;
        this.roomnumber = roomnumber;
        this.number = number;
        this.street = street;
        this.city = city;
        this.country = country;
        this.latitude = latitude;
        this.longitude = longitude;
        this.themes = themes;
        this.reviews = reviews;
        this.isOfficial = isOfficial;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDifficulty() {
        return difficulty;
    }

    public void setDifficulty(String difficulty) {
        this.difficulty = difficulty;
    }

    public double getMinprice() {
        return minprice;
    }

    public void setMinprice(double minprice) {
        this.minprice = minprice;
    }

    public double getMaxprice() {
        return maxprice;
    }

    public void setMaxprice(double maxprice) {
        this.maxprice = maxprice;
    }

    public String getImgurl() {
        return imgurl;
    }

    public void setImgurl(String imgurl) {
        this.imgurl = imgurl;
    }

    public String getWebsiteurl() {
        return websiteurl;
    }

    public void setWebsiteurl(String websiteurl) {
        this.websiteurl = websiteurl;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public int getMinplayer() {
        return minplayer;
    }

    public void setMinplayer(int minplayer) {
        this.minplayer = minplayer;
    }

    public int getMaxplayer() {
        return maxplayer;
    }

    public void setMaxplayer(int maxplayer) {
        this.maxplayer = maxplayer;
    }

    public int getRoomnumber() {
        return roomnumber;
    }

    public void setRoomnumber(int roomnumber) {
        this.roomnumber = roomnumber;
    }

    public int getNumber() {
        return number;
    }

    public void setNumber(int number) {
        this.number = number;
    }

    public String getStreet() {
        return street;
    }

    public void setStreet(String street) {
        this.street = street;
    }

    public String getCity() {
        return city;
    }

    public void setCity(String city) {
        this.city = city;
    }

    public String getCountry() {
        return country;
    }

    public void setCountry(String country) {
        this.country = country;
    }

    public double getLatitude() {
        return latitude;
    }

    public void setLatitude(double latitude) {
        this.latitude = latitude;
    }

    public double getLongitude() {
        return longitude;
    }

    public void setLongitude(double longitude) {
        this.longitude = longitude;
    }

    public List<String> getThemes() {
        return themes;
    }

    public void setThemes(List<String> themes) {
        this.themes = themes;
    }

    public List<Review> getReviews() {
        return reviews;
    }

    public void setReviews(List<Review> reviews) {
        this.reviews = reviews;
    }

    public Boolean getOfficial() {
        return isOfficial;
    }

    public void setOfficial(Boolean official) {
        isOfficial = official;
    }
}
